File: lib/formats/arc/ykc_archive.rb
require 'lib/formats/gfx/ykg_converter'
require 'lib/formats/script/yks_converter'
require 'lib/memory_file'

# YKC archive
# Engine: YukaScript
# Extension: .ykc
# Known games:
# - Hoshizora e Kakaru Hashi
module YkcArchive
  module_function

  MAGIC = 'YKC001'

  def add_cli_help(arg_parser)
    YkgConverter.add_cli_help(arg_parser)
    YksConverter.add_cli_help(arg_parser)
  end

  def parse_cli_options(arg_parser, options)
    YkgConverter.parse_cli_options(arg_parser, options)
    YksConverter.parse_cli_options(arg_parser, options)
  end

  class Unpacker
    def unpack(arc_file, output_files, options)
      magic = arc_file.read(MAGIC.length)
      fail RecognitionError, 'Not a YKC archive' unless magic == MAGIC

      _version,
      table_origin,
      table_size = arc_file.read(18).unpack('xxLx4L2')

      arc_file.seek(table_origin)
      num_files = table_size / 20

      num_files.times do
        output_files.write do
          name_origin,
          name_size,
          data_origin,
          data_size = arc_file.read(20).unpack('L4 x4')

          name = arc_file.peek(name_origin) { arc_file.read(name_size - 1) }
          data = arc_file.peek(data_origin) { arc_file.read(data_size) }

          file = MemoryFile.new(name, data)
          decode!(file, options)
          file
        end
      end
    end

    private

    def decode!(file, options)
      if file.data.start_with?(YkgConverter::MAGIC)
        file.data = YkgConverter.decode(file.data, options)
      elsif file.data.start_with?(YksConverter::MAGIC)
        file.data = YksConverter.decode(file.data, options)
      end
    end
  end

  class Packer
    def pack(arc_file, input_files, options)
      table_origin = MAGIC.length + 18
      table_size = input_files.length * 20

      version = 24
      arc_file.write(MAGIC)
      arc_file.write([version, table_origin, table_size].pack('xxLx4L2'))
      arc_file.write("\x00" * table_size)

      table_entries = {}
      input_files.each do |file|
        encode!(file, options)

        table_entries[file.name] = { name_origin: arc_file.tell }
        arc_file.write(file.name.gsub('/', '\\'))
        arc_file.write("\x00")

        table_entries[file.name][:data_origin] = arc_file.tell
        table_entries[file.name][:data_size] = file.data.length
        arc_file.write(file.data)
      end

      arc_file.seek(table_origin)
      table_entries.each do |name, entry|
        arc_file.write([
          entry[:name_origin],
          name.length + 1,
          entry[:data_origin],
          entry[:data_size]].pack('L4 x4'))
      end
    end

    def encode!(file, options)
      if file.name.downcase.end_with?('.ykg')
        file.data = YkgConverter.encode(file.data, options)
      elsif file.name.downcase.end_with?('.yks')
        file.data = YksConverter.encode(file.data, options)
      end
    end
  end
end

File: src/formats/arc/ykc_archive.cc
// YKC archive
//
// Company:   -
// Engine:    YukaScript
// Extension: .ykc
//
// Known games:
// - Hoshizora e Kakaru Hashi

#include "formats/arc/ykc_archive.h"
#include "formats/gfx/ykg_converter.h"

namespace
{
    const std::string magic("YKC001", 6);

    typedef struct
    {
        std::string name;
        size_t size;
        size_t offset;
    } TableEntry;

    typedef std::vector<std::unique_ptr<TableEntry>> Table;

    Table read_table(IO &arc_io, size_t table_offset, size_t table_size)
    {
        Table table;
        size_t file_count = table_size / 20;

        for (size_t i = 0; i < file_count; i ++)
        {
            std::unique_ptr<TableEntry> table_entry(new TableEntry);

            arc_io.seek(table_offset + i * 20);
            size_t name_origin = arc_io.read_u32_le();
            size_t name_size = arc_io.read_u32_le();
            table_entry->offset = arc_io.read_u32_le();
            table_entry->size = arc_io.read_u32_le();
            arc_io.skip(4);

            arc_io.seek(name_origin);
            table_entry->name = arc_io.read(name_size);
            table.push_back(std::move(table_entry));
        }

        return table;
    }

    std::unique_ptr<VirtualFile> read_file(
        IO &arc_io,
        TableEntry &table_entry,
        YkgConverter &ykg_converter)
    {
        std::unique_ptr<VirtualFile> file(new VirtualFile);
        arc_io.seek(table_entry.offset);
        file->io.write_from_io(arc_io, table_entry.size);
        file->name = table_entry.name;

        ykg_converter.try_decode(*file);
        // todo: decode with YKS once implemented

        return file;
    }
}

struct YkcArchive::Internals
{
    YkgConverter ykg_converter;
    // todo: YKS converters goes here
};

YkcArchive::YkcArchive() : internals(new Internals())
{
}

YkcArchive::~YkcArchive()
{
}

void YkcArchive::add_cli_help(ArgParser &arg_parser) const
{
    internals->ykg_converter.add_cli_help(arg_parser);
    // todo: pass to YKS once implemented
}

void YkcArchive::parse_cli_options(ArgParser &arg_parser)
{
    internals->ykg_converter.parse_cli_options(arg_parser);
    // todo: pass to YKS once implemented
}

void YkcArchive::unpack_internal(
    VirtualFile &file, OutputFiles &output_files) const
{
    if (file.io.read(magic.size()) != magic)
        throw std::runtime_error("Not a YKC archive");

    file.io.skip(2);
    int version = file.io.read_u32_le();
    file.io.skip(4);

    size_t table_offset = file.io.read_u32_le();
    size_t table_size = file.io.read_u32_le();
    Table table = read_table(file.io, table_offset, table_size);

    for (auto &table_entry : table)
    {
        output_files.save([&]() -> std::unique_ptr<VirtualFile>
        {
            return read_file(
                file.io,
                *table_entry,
                internals->ykg_converter);
        });
    }
}
